### PR TITLE
[LIVE-1174] - Feature: Upgrade NFT Architecture LLM V3

### PR DESCRIPTION
**Pull request by lambertkevin**
#### Dependency
https://github.com/LedgerHQ/ledger-live-common/pull/1938

### Type

Feature

### Context

Refactoring needing for the new NFT architecture + switch to TS for NFT related components

### Parts of the app affected / Test plan

Non reg on ETH NFTs

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -70,11 +70,11 @@
...
     "@formatjs/intl-numberformat": "^7.4.2",
     "@formatjs/intl-pluralrules": "^4.3.2",
-    "@ledgerhq/devices": "6.24.1",
+    "@ledgerhq/devices": "6.27.1",
     "@ledgerhq/errors": "6.10.0",
-    "@ledgerhq/hw-transport": "6.24.1",
-    "@ledgerhq/hw-transport-http": "6.27.0",
-    "@ledgerhq/live-common": "22.0.2",
+    "@ledgerhq/hw-transport": "6.27.1",
+    "@ledgerhq/hw-transport-http": "6.27.1",
```
Comment (haammar-ledger) [inline package.json:76]: Is it relevant to bump these here? Are you actually using the new versions? (excepted live-common)

File: src/components/Nft/NftCollectionRow.tsx [hunk 2]
```diff
@@ -41,7 +42,7 @@ function NftCollectionRow({ collection, onCollectionPress }: Props) {
               ellipsizeMode="tail"
               numberOfLines={2}
             >
-              {metadata?.tokenName || collection.contract}
+              {metadata?.tokenName || nft.contract}
```
Comment (haammar-ledger) [inline src/components/Nft/NftCollectionRow.tsx:45]: No need for a safety `nft?.contract` here?

File: src/components/Nft/NftCollectionRow.tsx [hunk 3]
```diff
@@ -51,14 +52,14 @@ function NftCollectionRow({ collection, onCollectionPress }: Props) {
           color={"neutral.c70"}
           ml={5}
         >
-          {collection.nfts.length}
+          {collection.length}
```
Comment (haammar-ledger) [inline src/components/Nft/NftCollectionRow.tsx:55]: Bascially, don't we want to `?` everything? :smile:

**Review by ghost — COMMENTED** [review]
Quick pass with 1st feedbacks

File: package.json (hunk 1 above)
Comment (lambertkevin) [reply package.json:76]: It's a dedup cause the build refused to work without it...

**Review by ghost — COMMENTED** [review]
2nd and last pass.
Mostly flagging potential optional chaining + a dumb question.

**Review by ghost — CHANGES_REQUESTED** [review]
A couple lint issues in the CI.

**Comment by lambertkevin** [comment]
> A couple lint issues in the CI.

Problem is I can't fix most of them, cause we're still running flow check on .tsx files 😞

File: src/components/Nft/NftCollectionRow.tsx (hunk 2 above)
Comment (lambertkevin) [reply src/components/Nft/NftCollectionRow.tsx]: Done 👌

File: src/components/Nft/NftCollectionRow.tsx (hunk 3 above)
Comment (lambertkevin) [reply src/components/Nft/NftCollectionRow.tsx]: Done 👌

File: src/components/Nft/NftViewer.tsx [hunk 4]
```diff
@@ -113,10 +114,12 @@ const NftViewer = ({ route }: Props) => {
...
 
+  const nftCapabilities = useMemo(() => getNftCapabilities(nft), [nft]);
+
   const defaultLinks = {
-    opensea: `https://opensea.io/assets/${collection.contract}/${nft.tokenId}`,
-    rarible: `https://rarible.com/token/${collection.contract}:${nft.tokenId}`,
-    etherscan: `https://etherscan.io/token/${collection.contract}?a=${nft.tokenId}`,
+    opensea: `https://opensea.io/assets/${nft.contract}/${nft.tokenId}`,
+    rarible: `https://rarible.com/token/${nft.contract}:${nft.tokenId}`,
+    etherscan: `https://etherscan.io/token/${nft.contract}?a=${nft.tokenId}`,
```
[the remark this answers is not in the document]
Comment (lambertkevin) [reply src/components/Nft/NftViewer.tsx]: Done 👌

File: src/components/Nft/NftViewer.tsx [hunk 5]
```diff
@@ -129,9 +132,10 @@ const NftViewer = ({ route }: Props) => {
     let transaction = bridge.createTransaction(account);
     transaction = bridge.updateTransaction(transaction, {
       tokenIds: [nft.tokenId],
-      quantities: [BigNumber(1)],
-      collection: collection.contract,
-      mode: `${collection.standard?.toLowerCase()}.transfer`,
+      // Quantity is set to null first to allow the user to change it on the amount page
+      quantities: [nftCapabilities.hasQuantity ? null : BigNumber(1)],
+      collection: nft.contract,
+      mode: `${nft.standard?.toLowerCase()}.transfer`,
```
[the remark this answers is not in the document]
Comment (lambertkevin) [reply src/components/Nft/NftViewer.tsx]: Done 👌

File: src/components/Nft/NftViewer.tsx [hunk 6]
```diff
@@ -306,7 +310,7 @@ const NftViewer = ({ route }: Props) => {
 
           <Section
             title={t("nft.viewer.tokenContract")}
-            value={collection.contract}
+            value={nft.contract}
```
[the remark this answers is not in the document]
Comment (lambertkevin) [reply src/components/Nft/NftViewer.tsx]: Done 👌

File: src/components/Nft/NftViewer.tsx [hunk 7]
```diff
@@ -320,7 +324,7 @@ const NftViewer = ({ route }: Props) => {
             copiedString={t("nft.viewer.tokenIdCopied")}
           />
 
-          {collection.standard === "ERC1155" && (
+          {nft.standard === "ERC1155" && (
```
[the remark this answers is not in the document]
Comment (lambertkevin) [reply src/components/Nft/NftViewer.tsx]: Done 👌

File: src/screens/Nft/NftCollection/NftCollectionHeaderTitle.tsx [hunk 8]
```diff
@@ -10,9 +9,12 @@ import LText from "../../../components/LText";
...
   const { params } = useRoute();
   const { colors } = useTheme();
+  const { collection } = params;
+  const nft = collection[0];
   const { status, metadata } = useNftMetadata(
-    params.collection.contract,
-    params.collection.nfts[0].tokenId,
+    nft.contract,
+    nft.tokenId,
+    nft.currencyId,
```
[the remark this answers is not in the document]
Comment (lambertkevin) [reply src/screens/Nft/NftCollection/NftCollectionHeaderTitle.tsx]: Done 👌

File: src/screens/Nft/NftCollection/NftCollectionHeaderTitle.tsx [hunk 9]
```diff
@@ -39,7 +41,7 @@ const NftCollectionHeaderTitle = () => {
           numberOfLines={1}
           style={styles.title}
         >
-          {metadata?.tokenName || params.collection.contract}
+          {metadata?.tokenName || collection.contract}
```
[the remark this answers is not in the document]
Comment (lambertkevin) [reply src/screens/Nft/NftCollection/NftCollectionHeaderTitle.tsx]: Done 👌

File: src/screens/Nft/NftGallery/NftCollectionWithName.tsx [hunk 10]
```diff
@@ -46,12 +41,12 @@ const NftCollectionWithNameList = ({
             semiBold
             style={styles.tokenName}
           >
-            {metadata?.tokenName || contract}
+            {metadata?.tokenName || nft.contract}
```
[the remark this answers is not in the document]
Comment (lambertkevin) [reply src/screens/Nft/NftGallery/NftCollectionWithName.tsx]: Done 👌

File: src/screens/Nft/NftGallery/NftCollectionWithName.tsx [hunk 11]
```diff
@@ -65,28 +60,32 @@ const NftCollectionWithNameMemo = memo(NftCollectionWithNameList);
...
+  collection,
   contentContainerStyle,
 }: Props) => {
-  const { contract, nfts } = collectionWithNfts;
-  const { status, metadata } = useNftMetadata(contract, nfts?.[0]?.tokenId);
+  const nft = collection[0];
+  const { status, metadata } = useNftMetadata(
+    nft.contract,
+    nft.tokenId,
+    nft.currencyId,
```
[the remark this answers is not in the document]
Comment (lambertkevin) [reply src/screens/Nft/NftGallery/NftCollectionWithName.tsx]: Done 👌

File: src/screens/SendFunds/01b-SelectCollection.tsx [hunk 12]
```diff
@@ -26,22 +26,18 @@ const MAX_COLLECTIONS_FIRST_RENDER = 8;
...
+  ({ account, collection }: { account: Account; collection: ProtoNFT[] }) => {
     const navigation = useNavigation();
     const { colors } = useTheme();
+    const nft = collection[0];
     const { status, metadata } = useNftMetadata(
-      collection?.contract,
-      collection?.nfts?.[0]?.tokenId,
+      nft.contract,
+      nft.tokenId,
+      nft.currencyId,
```
[the remark this answers is not in the document]
Comment (lambertkevin) [reply src/screens/SendFunds/01b-SelectCollection.tsx]: Done 👌

File: src/screens/SendFunds/01b-SelectCollection.tsx [hunk 13]
```diff
@@ -61,7 +57,7 @@ const CollectionRow = memo(
             style={[styles.tokenNameSkeleton, styles.tokenName]}
             loading={status === "loading"}
           >
-            <LText>{metadata?.tokenName || collection.contract}</LText>
+            <LText>{metadata?.tokenName || nft.contract}</LText>
```
[the remark this answers is not in the document]
Comment (lambertkevin) [reply src/screens/SendFunds/01b-SelectCollection.tsx]: Done 👌

File: src/screens/SendFunds/01c-SelectNft.tsx [hunk 14]
```diff
@@ -26,93 +25,87 @@ import { ScreenName } from "../../const";
...
-              {metadata?.nftName || "-"}
-            </LText>
-          </Skeleton>
+const NftRow = memo(({ account, nft }: { account: Account; nft: ProtoNFT }) => {
+  const navigation = useNavigation();
+  const { colors } = useTheme();
+  const { status, metadata } = useNftMetadata(
+    nft.contract,
+    nft.tokenId,
+    nft.currencyId,
```
[the remark this answers is not in the document]
Comment (lambertkevin) [reply src/screens/SendFunds/01c-SelectNft.tsx]: Done 👌

File: src/screens/SendFunds/01c-SelectNft.tsx [hunk 15]
```diff
@@ -26,93 +25,87 @@ import { ScreenName } from "../../const";
...
+  const goToRecipientSelection = useCallback(() => {
+    const bridge = getAccountBridge(account);
+
+    let transaction = bridge.createTransaction(account);
+    transaction = bridge.updateTransaction(transaction, {
+      tokenIds: [nft.tokenId],
+      // Quantity is set to null first to allow the user to change it on the amount page
+      quantities: [nftCapabilities.hasQuantity ? null : BigNumber(1)],
+      collection: nft.contract,
+      mode: `${nft.standard?.toLowerCase()}.transfer`,
```
[the remark this answers is not in the document]
Comment (lambertkevin) [reply src/screens/SendFunds/01c-SelectNft.tsx]: Done 👌

File: src/screens/SendFunds/01c-SelectNft.tsx [hunk 16]
```diff
@@ -26,93 +25,87 @@ import { ScreenName } from "../../const";
...
+          ellipsizeMode="middle"
+          style={[styles.tokenId, { color: colors.grey }]}
+        >
+          ID {nft.tokenId}
+        </LText>
+      </View>
+      {nft?.standard === "ERC1155" ? (
+        <View style={styles.amount}>
+          <LText numberOfLines={1} style={{ color: colors.grey }}>
+            x{nft.amount.toFixed()}
```
[the remark this answers is not in the document]
Comment (lambertkevin) [reply src/screens/SendFunds/01c-SelectNft.tsx]: Done 👌

File: src/screens/SendFunds/03b-AmountNft.tsx [hunk 17]
```diff
@@ -71,22 +70,18 @@ export default function SendAmountNFT({ route }: Props) {
...
-  const nft = account?.nfts?.find(
-    nft =>
-      nft.collection.contract === transaction?.collection &&
-      nft.tokenId === transaction?.tokenIds[0],
+  const nft = useMemo(
+    () =>
+      account?.nfts?.find(
+        nft =>
+          nft.contract === transaction?.collection &&
+          nft.tokenId === transaction?.tokenIds[0],
```
[the remark this answers is not in the document]
Comment (lambertkevin) [reply src/screens/SendFunds/03b-AmountNft.tsx]: Done 👌

File: src/components/Skeleton.tsx [hunk 18]
```diff
@@ -1,20 +1,18 @@
...
+  loading: boolean;
+  children?: ReactElement;
+  animated?: boolean;
 };
 
 const Skeleton = ({
   style,
   loading,
-  children = null,
+  children = <></>,
```
[the remark this answers is not in the document]
Comment (haammar-ledger) [reply src/components/Skeleton.tsx:15]: LGTM!

**Review by ghost — APPROVED** [review]

**Comment by haammar-ledger** [comment]
(Lint errors in CI are flow errors on TS files...)

**Pull request merged**